New post: custom bullets with marker (#3580)
GoogleChrome/web.dev: src/site/content/en/blog/css-marker-pseudo-element/index.md

---

> title: Custom Bullets with CSS ::marker

Reviewer: ```suggestion
title: Custom bullets with CSS ::marker
```
Our title convention is sentence case

subhead: It is now trivial to customize the color, size or type of number or bullet when using a <ul> or <ol>. 
authors:
  - adamargyle
  - loirooriol

> description: It is now trivial to customize the color, size or type of number or bullet when using a <ul> or <ol>.

Reviewer: ```suggestion
description: It is now trivial to customize the color, size or type of number or bullet when using a <ul> or <ol>.
tags:
  - blog
  - css
```
We need `blog` for this to show up on web.dev/blog

Other relevant tags can be found at https://github.com/GoogleChrome/web.dev/blob/master/src/site/_data/tagsData.json

FYI @rachelandrew

date: 2020-07-27

> updated: 2020-07-27

Reviewer: ```suggestion
#updated: 2020-07-27
```
I usually omit this until we actually update the content

> hero: hero.jpg

Reviewer: The hero image appears to be a 404

> alt: TODO

Reviewer: Need to provide an `alt` for the hero image

--- 

Thanks to Igalia, sponsored by Bloomberg, we can finally put our hacks away for styling lists. See!

<figure class="w-figure">
  <img class="w-screenshot" src="marker-fun.png">
  <figcaption class="w-figcaption">
    <a href="https://glitch.com/edit/#!/marker-fun-example">View Source</a>
  </figcaption>
</figure>

> Thanks to [CSS `::marker`](https://www.w3.org/TR/css-lists-3/#marker-pseudo) we can change the content and some of the styles of bullets and numbers. At the time of writing `::marker` is [supported](https://caniuse.com/#search=%3A%3Amarker) in Firefox 68 for desktop and Android, Safari 11.1 and iOS Safari 11.3, and Chromium 86 (desktop and Android).
> 
> ## Pseudo-elements

Reviewer: ```suggestion
Thanks to [CSS `::marker`](https://www.w3.org/TR/css-lists-3/#marker-pseudo) we can change the content and some of the styles of bullets and numbers. 

## Browser compatibilty

At the time of writing `::marker` is supported in Firefox for desktop and Android, desktop Safari and iOS Safari, and Chromium-based desktop and Android browsers. See MDN's [Browser compatibility](https://developer.mozilla.org/en-US/docs/Web/CSS/::marker#Browser_compatibility) table for updates.

## Pseudo-elements
```

Are you OK with me adding a section heading? I like to make this information very easy to find. I'm also avoiding mentioning specific versions of the browsers. I think the important thing is that it's supported in these browsers. I don't think people need to know the exact versions. If they do, they can go to MDN's table. I also prefer to link to MDN rather than caniuse for consistency. They both pull from the same source but MDN is more official.

</div>

The dot at the beginning of each `<li>` item is free! The browser is drawing and creating a generated marker box for you. 

Today we're excited to talk about the `::marker` pseudo element, which gives the ability to style the bullet element that browsers create for you.

{% Aside 'key-term' %}
A pseudo-element represents an element in the document other than those which exist in the document tree. For example, you can select the first line of a paragraph using the pseudo-element `p::first-line`, even though there is no HTML element wrapping that line of text.
{% endAside %}

### Creating a marker

The `::marker` pseudo-element marker box is automatically generated inside every list item element, preceding the actual contents and the `::before` pseudo-element.

```css
li::before {
  content: "::before";
  background: lightgray;
  border-radius: 1rem;
  padding-inline: 1ch;
  margin-inline-end: 1ch;
}
```

<div class="glitch-embed-wrap" style="height: 340px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/marker-before-example?path=style.css&previewSize=100"
    alt="List Demo on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>

> </div>

Reviewer: Fun demo :D

![image](https://user-images.githubusercontent.com/4713486/88848380-3d7e9d00-d19d-11ea-8fda-664596d430f9.png)

Also, note how the disallowed `background` property has no effect.

<div class="w-columns">
{% Compare 'worse', 'List Styles' %}
```css
li:nth-child(1) {
  list-style-type: '?';
  font-size: 2rem;
  background: hsl(200 20% 88%);
  animation: color-change 3s ease-in-out infinite;
}
```

{% CompareCaption %}
Mixed results between the marker and the list item
{% endCompareCaption %}

{% endCompare %}

{% Compare 'better', 'Marker Styles' %}
```css
li:nth-child(2)::marker {
  content: '!';
  font-size: 2rem;
  background: hsl(200 20% 88%);
  animation: color-change 3s ease-in-out infinite;
}
```

{% CompareCaption %}
Focused results between marker and list item
{% endCompareCaption %}

{% endCompare %}
</div>

<br>

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/marker-style-vs-list-style-type?path=style.css&previewSize=100"
    alt="List Demo on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

> {% Aside 'gotchas' %}

Author: tbh this whole gotcha confuses me, can we get rid of it entirely and leave it for spec implementation details in other articles?

In Chromium, `white-space` only works for inside positioned markers. For outside positioned markers, the style adjuster always forces `white-space: pre` in order to preserve the trailing space.
{% endAside %}

#### Changing the content of a marker
This CSS property can be used in a few ways and we felt it was important to cover a few. 

**Changing all list items**
```css
li {
  list-style-type: "😍";
}

/* OR */

li::marker {
  content: "😍";
}
```

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/marker-change-all?path=style.css&previewSize=100"
    alt="List Demo on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

> **Changing just 1 list item**

Reviewer: ```suggestion
**Changing just one list item**
```
I just rediscovered that our style guide tells us to spell out one through nine

}
```

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/marker-inline-svg?path=style.css&previewSize=100"
    alt="List Demo on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

**Changing numbered lists**  
What about an `<ol>` though right? Each of those `::markers` aren't the simple bullet, they're numbers. In CSS these are called [Counters](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Lists_and_Counters/Using_CSS_counters), and they're quite powerful. They even have properties to set and reset where the number starts and ends, or switching them to roman numerals. Can we style that? Yep, and we can even use the marker content value to build our own numbering presentation.

```css
li::marker {
  content: counter(list-item) "› "; 
  color: hotpink;
}
```

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/marker-numbered-lists?path=style.css&previewSize=100"
    alt="List Demo on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>

> </div>

Reviewer: It might be worthwhile to add some feature detection or warning text to the demo saying that this won't work unless you're on the correct version of Chrome, because this is what I saw as I scanned through the content

![image](https://user-images.githubusercontent.com/4713486/88848800-ce557880-d19d-11ea-8b69-8774d797d75d.png)

One might argue that the reader should figure this out themselves, but I'm a fan of preventing confusion as much as possible

Author: good call, i've added this, 

![image](https://user-images.githubusercontent.com/1134620/89076930-17403500-d336-11ea-8fc5-131e40d2ce5b.png)

### Debugging

> Chromium devtools is ready to help you inspect, debug and modify the styles applying to `::marker` pseudo elements.

Reviewer 2: ```suggestion
Chrome DevTools is ready to help you inspect, debug and modify the styles applying to `::marker` pseudo elements.
```

Reviewer: I'm OK with Chromium DevTools or Chrome DevTools but DevTools should be capitalized the way that Rachel and I am doing it (to keep the branding consistent across our other docs)

<figure class="w-figure">
  <img class="w-screenshot" style="max-inline-size: 480px" src='./devtools.png' alt="DevTools open and showing styles from the user agent and the user styles">
</figure>

> ### Future Pseudo Element Styling

Reviewer: ```suggestion
### Future Pseudo-element styling
```
We use sentence case for our section headings

---
title: Custom bullets with CSS ::marker
subhead: It is now trivial to customize the color, size or type of number or bullet when using a <ul> or <ol>. 
authors:
  - adamargyle
  - loirooriol
description: It is now trivial to customize the color, size or type of number or bullet when using a <ul> or <ol>.
tags:
  - blog
  - css

> date: 2020-07-27

Reviewer: ```suggestion
date: 2020-08-05
```
